### PR TITLE
[3006.x] Remove the use of `curl_httpclient` from `salt.utils.http`

### PR DESCRIPTION
**Pull request by MKLeb**
### What does this PR do?
See title

### What issues does this PR fix or reference?
Fixes: https://github.com/saltstack/salt/issues/64371

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Comment by Ch3LL** [comment]
This seems like it should be only added to hte `master` branch. Also, shouldn't we put it on a deprecation path first before removing?

**Comment by MKLeb** [comment]
> This seems like it should be only added to hte `master` branch. Also, shouldn't we put it on a deprecation path first before removing?

A valid question.  @dwoz?
From my understanding, we are simply forcing the use of proxies to go through the requests backend, and we already package with requests so it should be there.

**Comment by lukasraska** [comment]
@MKLeb For what it's worth, you're forcing usage of requests if proxies are used, so that behaviour is different from when tornado is being used (those two backends DO behave differently in certain situations)

This is then in direct contradiction with Salt Project LTS/STS release and support policy - https://docs.saltproject.io/salt/install-guide/en/latest/topics/salt-version-support-lifecycle.html , which clearly states `The purpose of the LTS release is to provide users with a stable version of Salt for a longer period (2 years).` and is best for users who:

- Need or prefer a stable, thoroughly tested and well-supported release
- Prefer stable releases over access to the latest features
- Require longer time periods to plan and implement Salt infrastructure upgrades


However removing / changing usage of module in bugfix version is not something I would call stable. Although pycurl doesn't seem to be included in Onedir 3006 packages, it can still be used and **tornado is the default backend** - so it can really cause some issues if this is not tested thorougly

File: salt/utils/http.py [hunk 1]
```diff
@@ -192,6 +185,36 @@ def query(
...
+    no_proxy = opts.get("no_proxy", [])
+
+    if urllib.parse.urlparse(url).hostname in no_proxy:
+        proxy_host = None
+        proxy_port = None
+        proxy_username = None
+        proxy_password = None
+
+    http_proxy_url = None
+    if backend != "requests" and proxy_host and proxy_port:
```
Comment (twangboy) [inline salt/utils/http.py:207]: Could you move this if statement under the following if statement? Then it would just be
```
if backend != "requests":
```

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/utils/http.py (hunk 1 above)
Comment (MKLeb) [reply salt/utils/http.py:207]: yup, good catch

**Review by s0undt3ch — DISMISSED** [review]

File: tests/pytests/unit/utils/test_http.py [hunk 2]
```diff
@@ -49,5 +53,263 @@ def test_session_ca_bundle():
...
+    """
+    This tests that we get a null response when raise_error=False and the
+    host/port cannot be reached.
+    """
+    host = "127.0.0.1"
+
+    # Find unused port
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        sock.bind((host, 0))
+        port = sock.getsockname()[1]
```
Comment (s0undt3ch) [inline tests/pytests/unit/utils/test_http.py:156]:
```suggestion
    port = ports.get_unused_localhost_port()
```

And, at the top
```python
from pytestshellutils.utils import ports
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by s0undt3ch — DISMISSED** [review]

**Pull request merged**